File: backend/app/schemas.py
from pydantic import BaseModel, EmailStr
from datetime import datetime
from typing import List, Optional

# User Schemas
class UserBase(BaseModel):
    username: str
    email: EmailStr

class UserCreate(UserBase):
    password: str

class UserLogin(BaseModel):
    username: str
    password: str

class UserResponse(UserBase):
    id: int
    role: str
    is_active: bool
    created_at: datetime

    class Config:
        from_attributes = True

# Question Schemas
class QuestionBase(BaseModel):
    title: str
    description: str
    tags: List[str] = []

class QuestionCreate(QuestionBase):
    pass

class QuestionResponse(QuestionBase):
    id: int
    owner_id: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True

# Answer Schemas
class AnswerBase(BaseModel):
    content: str

class AnswerCreate(AnswerBase):
    question_id: int

class AnswerResponse(AnswerBase):
    id: int
    owner_id: int
    question_id: int
    is_accepted: bool
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True

# Tag Schemas
class TagBase(BaseModel):
    name: str

class TagResponse(TagBase):
    id: int

    class Config:
        from_attributes = True

# Vote Schemas
class VoteCreate(BaseModel):
    answer_id: int
    type: int # 1 for upvote, -1 for downvote

class VoteResponse(VoteCreate):
    user_id: int

    class Config:
        from_attributes = True

# Notification Schemas
class NotificationResponse(BaseModel):
    id: int
    user_id: int
    message: str
    is_read: bool
    created_at: datetime

    class Config:
        from_attributes = True

File: backend/schemas/schemas.go
package schemas

import "time"

// User Schemas
type UserCreate struct {
	Username string `json:"username" validate:"required"`
	Email    string `json:"email" validate:"required,email"`
	Password string `json:"password" validate:"required,min=6"`
}

type UserLogin struct {
	Username string `json:"username" validate:"required"`
	Password string `json:"password" validate:"required"`
}

type UserResponse struct {
	ID        uint      `json:"id"`
	Username  string    `json:"username"`
	Email     string    `json:"email"`
	Role      string    `json:"role"`
	IsActive  bool      `json:"is_active"`
	CreatedAt time.Time `json:"created_at"`
}

type Token struct {
	AccessToken string `json:"access_token"`
	TokenType   string `json:"token_type"`
}

// Question Schemas
type QuestionCreate struct {
	Title       string   `json:"title" validate:"required"`
	Description string   `json:"description" validate:"required"`
	Tags        []string `json:"tags"`
}

type QuestionResponse struct {
	ID          uint          `json:"id"`
	Title       string        `json:"title"`
	Description string        `json:"description"`
	OwnerID     uint          `json:"owner_id"`
	Tags        []TagResponse `json:"tags"` // Include tags in the response
	CreatedAt   time.Time     `json:"created_at"`
	UpdatedAt   *time.Time    `json:"updated_at,omitempty"`
}

// Answer Schemas
type AnswerCreate struct {
	Content    string `json:"content" validate:"required"`
	QuestionID uint   `json:"question_id" validate:"required"`
}

type AnswerResponse struct {
	ID         uint       `json:"id"`
	Content    string     `json:"content"`
	QuestionID uint       `json:"question_id"`
	OwnerID    uint       `json:"owner_id"`
	IsAccepted bool       `json:"is_accepted"`
	CreatedAt  time.Time  `json:"created_at"`
	UpdatedAt  *time.Time `json:"updated_at,omitempty"`
}

// Tag Schemas
type TagResponse struct {
	ID   uint   `json:"id"`
	Name string `json:"name"`
}

// Vote Schemas
type VoteCreate struct {
	AnswerID uint `json:"answer_id" validate:"required"`
	Type     int  `json:"type" validate:"required,oneof=1 -1"` // 1 for upvote, -1 for downvote
}

// Notification Schemas
type NotificationResponse struct {
	ID        uint      `json:"id"`
	UserID    uint      `json:"user_id"`
	Message   string    `json:"message"`
	IsRead    bool      `json:"is_read"`
	CreatedAt time.Time `json:"created_at"`
}
